tools/hdr/local_tone_map.cc

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "lib/extras/codec.h"
+#include "lib/extras/tone_mapping.h"
+#include "lib/jxl/convolve.h"
+#include "lib/jxl/enc_color_management.h"
+#include "lib/jxl/enc_gamma_correct.h"
+#include "lib/jxl/image_bundle.h"
+#include "tools/args.h"
+#include "tools/cmdline.h"
+#include "tools/thread_pool_internal.h"
+
+namespace jxl {
+namespace {
+
+constexpr WeightsSeparable5 kPyramidFilter = {
+    {HWY_REP4(.375f), HWY_REP4(.25f), HWY_REP4(.0625f)},
+    {HWY_REP4(.375f), HWY_REP4(.25f), HWY_REP4(.0625f)}};
+
+// Expects sRGB input.
+// Will call consumer(x, y, contrast) for each pixel.
+template <typename Consumer>
+void Contrast(const jxl::Image3F& image, const Consumer& consumer,
```

Comment:
Ah, and the callback was so that it wouldn’t have to construct an intermediary `ImageF` – but since it’s always used to add (a weighted log) to an `ImageF`, I suppose it could take a pointer to it and do the calculation directly?

Reply:
that might be a good idea, but I wouldn't worry too much about it :)